File: Tang/js/fktv.js
const host = 'https://fktv.me';

const headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
};

// ========== 替换为系统原生标准 Base64（双平台100%兼容） ==========
function base64Encode(text) {
    try {
        return btoa(unescape(encodeURIComponent(text)));
    } catch (e) {
        return "";
    }
}

function base64Decode(str) {
    try {
        return decodeURIComponent(escape(atob(str)));
    } catch (e) {
        return "";
    }
}

function generateCookie() {
    const t = "ABCDEFGHJKMNPQRSTWXYZabcdefhijkmnprstwxyz102345678";
    let n = "";
    for (let i = 0; i < 32; i++) {
        n += t.charAt(Math.floor(Math.random() * t.length));
    }
    return `_did=${n}`;
}

function getProxyUrl(imgUrl) {
    if (!imgUrl) return "";
    const enc = base64Encode(encodeURIComponent(imgUrl));
    return `proxy://do=fktv_img&url=${enc}`;
}

// ========== 标准接口（Fongmi 严格兼容版） ==========
async function init(cfg) {}

async function home() {
    return JSON.stringify({
        class: [
            { type_id: '1', type_name: '电影' },
            { type_id: '2', type_name: '电视剧' },
            { type_id: '4', type_name: '动漫' },
            { type_id: '3', type_name: '综艺' },
            { type_id: '8', type_name: '短剧' },
            { type_id: '6', type_name: '纪录片' }
        ]
    });
}

async function homeVod() {
    return JSON.stringify({ list: [], page: 1, pagecount: 0 });
}

// ========== 【重点修复】分类列表（纯正则解析，绕开Fongmi解析器） ==========
async function category(tid, pg, filter, extend) {
    const page = Number(pg) || 1;
    const url = `${host}/channel?page=${page}&cat_id=${tid}&order=new&page_size=32`;
    try {
        const res = await req(url, { headers });
        const html = res.content || '';
        let list = [];

        // 用纯正则匹配所有 item-wrap，彻底绕开 pdfa 解析器
        const items = html.match(/<div class="item-wrap[\s\S]*?<\/div>\s*<\/div>/g);
        if (items && items.length > 0) {
            list = items.map(itHtml => {
                // 提取名称
                let nameMatch = itHtml.match(/<a[^>]*title="([^"]+)"/);
                let name = nameMatch ? nameMatch[1] : "";
                
                // 提取图片地址
                let picMatch = itHtml.match(/data-src="([^"]+)"/);
                let pic = picMatch ? getProxyUrl(picMatch[1]) : "";

                // 提取备注
                let remarkMatch = itHtml.match(/<div class="category[^>]*>([^<]+)<\/div>/);
                let remark = remarkMatch ? remarkMatch[1].trim() : "";

                // 提取 vod_id
                let idMatch = itHtml.match(/<a[^>]*href="([^"]+)"/);
                let id = idMatch ? idMatch[1] : "";

                // 【关键兜底】如果代理图片失效，用占位图，保证列表项能渲染出来
                if (!pic) {
                    pic = `https://via.placeholder.com/300x450/333/fff?text=${encodeURIComponent(name || "无图")}`;
                }

                return {
                    vod_name: name,
                    vod_pic: pic,
                    vod_remarks: remark,
                    vod_id: id
                };
            }).filter(item => item.vod_name && item.vod_id);
        }

        return JSON.stringify({
            page: page,
            pagecount: 99,
            list: list
        });
    } catch (e) {
        return JSON.stringify({ page: 1, pagecount: 0, list: [] });
    }
}

function extractLinesAndEpisodes(html, vodId) {
    // 提取线路名：匹配 data-line 属性对应的 div 内容
    const lineRegex = /<div[^>]*data-line="([^"]+)"[^>]*>([^<]+)<\/div>/g;
    let lines = [];
    let match;
    while ((match = lineRegex.exec(html)) !== null) {
        lines.push({ id: match[1], name: match[2].trim() });
    }
    // 如果没有 data-line，尝试匹配 .play-source
    if (lines.length === 0) {
        const altLineRegex = /<div[^>]*class="[^"]*play-source[^"]*"[^>]*>([^<]+)<\/div>/g;
        while ((match = altLineRegex.exec(html)) !== null) {
            lines.push({ id: `line${lines.length+1}`, name: match[1].trim() });
        }
    }

    // 提取所有集数：匹配 data-id 和集数名称
    const epRegex = /<div[^>]*data-id="([^"]+)"[^>]*>[\s\S]*?<span[^>]*class="[^"]*number[^"]*"[^>]*>([^<]+)<\/span>/g;
    let episodes = [];
    while ((match = epRegex.exec(html)) !== null) {
        episodes.push({ id: match[1], name: match[2].trim() });
    }

    // 如果集数没匹配到，尝试更宽松的匹配
    if (episodes.length === 0) {
        const altEpRegex = /<a[^>]*href="javascript:;"[^>]*data-id="([^"]+)"[^>]*>[\s\S]*?<span[^>]*>([^<]+)<\/span>/g;
        while ((match = altEpRegex.exec(html)) !== null) {
            episodes.push({ id: match[1], name: match[2].trim() });
        }
    }

    // 构建线路和集数字符串
    let playFroms = [];
    let playUrls = [];
    for (let line of lines) {
        playFroms.push(line.name);
        let epStr = episodes.map(ep => `${ep.name}$${line.id}-${vodId}-${ep.id}`).join('#');
        playUrls.push(epStr);
    }
    // 如果没有任何线路，至少给一个默认
    if (playFroms.length === 0 && episodes.length > 0) {
        playFroms.push('默认线路');
        let epStr = episodes.map(ep => `${ep.name}$${vodId}-${ep.id}`).join('#');
        playUrls.push(epStr);
    }

    return {
        vod_play_from: playFroms.join('$$$'),
        vod_play_url: playUrls.join('$$$')
    };
}

// ========== 【精准提取版】Detail 函数 ==========
async function detail(id) {
    const vodId = Array.isArray(id) ? id[0] : id;
    if (!vodId) return JSON.stringify({ list: [] });

    try {
        let url = vodId.startsWith('http') ? vodId : `${host}${vodId}`;
        const res = await req(url, { headers: headers, timeout: 10000 });
        const html = res.content || '';

        // --- 1. 提取基础信息 ---
        let vod_name = "未知影片";
        const nameMatch = html.match(/<h1[^>]*>([^<]+)<\/h1>/);
        if (nameMatch) vod_name = nameMatch[1].trim();

        let vod_pic = "";
        const picMatch = html.match(/data-src="([^"]+)"/) || html.match(/poster="([^"]+)"/);
        if (picMatch) vod_pic = getProxyUrl(picMatch[1]);
        else vod_pic = `https://via.placeholder.com/300x450/333/fff?text=${encodeURIComponent(vod_name)}`;

        let vod_content = "";
        const contentMatch = html.match(/class="desc"[^>]*>([\s\S]*?)<\/div>/);
        if (contentMatch) vod_content = contentMatch[1].replace(/<[^>]+>/g, '').trim();

        // --- 2. 调用核心解析函数（线路+集数）---
        const { vod_play_from, vod_play_url } = extractLinesAndEpisodes(html, vodId);

        // --- 3. 返回结果 ---
        return JSON.stringify({
            list: [{
                vod_id: vodId,
                vod_name: vod_name,
                vod_pic: vod_pic,
                vod_content: vod_content,
                vod_play_from: vod_play_from,
                vod_play_url: vod_play_url
            }]
        });

    } catch (err) {
        console.error("Detail Error:", err);
        return JSON.stringify({ list: [] });
    }
}

// ========== 【最终修复版】主动请求API获取m3u8地址 ==========
async function play(flag, id, flags) {
    // 重要：这里的 id 正是你在 detail 函数里构造的格式（例如 "line1-/movie/detail/xxx-6c5ad..."）
    // 第一步：解析出我们需要的关键信息
    let lineId, vodId, epId;
    const parts = id.split('$');
    // 你目前的 detail 函数构造的 id 可能没有 '$'，所以这里提供两种解析方式，确保兼容
    let identifier = parts.length > 1 ? parts[1] : id;
    const dashParts = identifier.split('-');
    
    if (dashParts.length >= 3) {
        // 标准格式：lineId-vodId-epId
        lineId = dashParts[0];
        vodId = dashParts[1];
        epId = dashParts[2];
    } else if (dashParts.length === 2) {
        // 备选格式：vodId-epId
        lineId = 'default';
        vodId = dashParts[0];
        epId = dashParts[1];
    } else {
        // 降级处理
        vodId = identifier;
        epId = '1';
        lineId = 'default';
    }

    // 第二步：请求播放详情页，获取真实的播放地址
    const detailUrl = `${host}${vodId}`;
    const body = `link_id=${epId}&is_switch=1`;

    try {
        // 用 POST 方法，带着正确的 Cookie 和 Referer 去请求
        const res = await req(detailUrl, {
            method: 'POST',
            headers: {
                "Content-Type": 'application/x-www-form-urlencoded; charset=UTF-8',
                "Referer": detailUrl,
                "Cookie": generateCookie(), // 别忘了你的Cookie生成函数
                "User-Agent": headers["User-Agent"]
            },
            body: body
        });
        
        const response = JSON.parse(res.content);
        let m3u8Url = null;
        
        // 第三步：从返回的 JSON 数据里提取 m3u8 地址
        if (response.data && response.data.play_links) {
            if (lineId !== 'default') {
                const link = response.data.play_links.find(i => i.id == lineId);
                if (link) m3u8Url = link.m3u8_url;
            } else if (response.data.play_links.length > 0) {
                m3u8Url = response.data.play_links[0].m3u8_url;
            }
        }
        
        // 第四步：成功拿到地址，直接返回给播放器
        if (m3u8Url) {
            const fullUrl = m3u8Url.startsWith('http') ? m3u8Url : `${host}${m3u8Url}`;
            return JSON.stringify({ parse: 0, url: fullUrl });
        } else {
            // 如果没拿到，退一步让播放器自己去页面上嗅探
            return JSON.stringify({ parse: 1, url: detailUrl });
        }
    } catch (e) {
        // 报错时也尝试用嗅探兜底
        return JSON.stringify({ parse: 1, url: detailUrl });
    }
}


// ========== 【重点修复】搜索函数（和分类一样改用纯正则） ==========
async function search(wd, quick, pg) {
    const page = Number(pg) || 1;
    const url = `${host}/channel?page=${page}&keywords=${encodeURIComponent(wd)}&page_size=32&order=new`;
    try {
        const res = await req(url, { headers });
        const html = res.content || '';
        let list = [];

        const items = html.match(/<div class="item-wrap[\s\S]*?<\/div>\s*<\/div>/g);
        if (items && items.length > 0) {
            list = items.map(itHtml => {
                let nameMatch = itHtml.match(/<a[^>]*title="([^"]+)"/);
                let name = nameMatch ? nameMatch[1] : "";
                
                let picMatch = itHtml.match(/data-src="([^"]+)"/);
                let pic = picMatch ? getProxyUrl(picMatch[1]) : "";

                let remarkMatch = itHtml.match(/<div class="category[^>]*>([^<]+)<\/div>/);
                let remark = remarkMatch ? remarkMatch[1].trim() : "";

                let idMatch = itHtml.match(/<a[^>]*href="([^"]+)"/);
                let id = idMatch ? idMatch[1] : "";

                if (!pic) {
                    pic = `https://via.placeholder.com/300x450/333/fff?text=${encodeURIComponent(name || "无图")}`;
                }

                return {
                    vod_name: name,
                    vod_pic: pic,
                    vod_remarks: remark,
                    vod_id: id
                };
            }).filter(item => item.vod_name && item.vod_id);
        }

        return JSON.stringify({
            page: page,
            pagecount: 99,
            list: list
        });
    } catch (e) {
        return JSON.stringify({ page: 1, pagecount: 0, list: [] });
    }
}

async function proxy(args) {
    const doWhat = args.do || "";
    if (doWhat === 'fktv_img') {
        const encUrl = args.url || "";
        const targetUrl = base64Decode(encUrl);
        if (targetUrl) {
            return ["redirect", targetUrl];
        }
    }
    return [404, "text/plain", "Not Found"];
}

export default { init, home, homeVod, category, detail, search, play, proxy };

File: Tang/py/fktv.py
# -*- coding: utf-8 -*-
import requests
import re
import random
from typing import Dict, List, Optional
import sys

sys.path.append('..')
from base.spider import Spider


class Spider(Spider):
    """凡客影视 Python 爬虫 - 完整重写版（严格遵循JS逻辑）"""

    def __init__(self):
        super().__init__()
        self.session = None
        self._init_config()

    def _init_config(self):
        """初始化配置"""
        self.home_url = 'https://fktv.me'
        self.timeout = 15
        self.headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            "Referer": self.home_url + '/'
        }

    # ----------------------------------------------
    # 辅助函数
    # ----------------------------------------------

    def _generate_cookie(self) -> str:
        """生成随机 _did Cookie（32位字符串）"""
        chars = "ABCDEFGHJKMNPQRSTWXYZabcdefhijkmnprstwxyz102345678"
        return ''.join(random.choice(chars) for _ in range(32))

    def _get_placeholder_pic(self, name: str) -> str:
        """生成占位图片URL"""
        return f"https://via.placeholder.com/300x450/333/fff?text={name}"

    def _make_request(self, url: str, method: str = 'GET', **kwargs) -> Optional[requests.Response]:
        """统一请求方法（带Referer和异常处理）"""
        try:
            if method.upper() == 'GET':
                resp = self.session.get(url, timeout=self.timeout, **kwargs)
            else:
                resp = self.session.post(url, timeout=self.timeout, **kwargs)
            resp.raise_for_status()
            return resp
        except Exception as e:
            print(f"请求失败: {url} | 错误: {str(e)}")
            return None

    # ----------------------------------------------
    # 核心逻辑（线路&集数提取）
    # ----------------------------------------------

    def _extract_lines_and_episodes(self, html: str, vod_id: str) -> Dict[str, str]:
        """
        完全复刻JS版 extractLinesAndEpisodes 逻辑
        返回: {
            'vod_play_from': '线路1$$$线路2',
            'vod_play_url': '1$line1-vodId-epId#2$line1...$$$1$line2...'
        }
        """
        # 1. 提取线路（优先data-line，次选play-source）
        lines = []
        line_pattern = r'<div[^>]*data-line="([^"]+)"[^>]*>([^<]+)</div>'
        for m in re.finditer(line_pattern, html):
            lines.append({'id': m.group(1), 'name': m.group(2).strip()})
        
        if not lines:
            alt_line_pattern = r'<div[^>]*class="[^"]*play-source[^"]*"[^>]*>([^<]+)</div>'
            for idx, name in enumerate(re.findall(alt_line_pattern, html), 1):
                lines.append({'id': f'line{idx}', 'name': name.strip()})

        # 2. 提取集数（优先data-id，次选a标签）
        episodes = []
        ep_pattern = r'<div[^>]*data-id="([^"]+)"[^>]*>.*?<span[^>]*class="[^"]*number[^"]*"[^>]*>([^<]+)</span>'
        for m in re.finditer(ep_pattern, html, re.DOTALL):
            episodes.append({'id': m.group(1), 'name': m.group(2).strip()})
        
        if not episodes:
            alt_ep_pattern = r'<a[^>]*data-id="([^"]+)"[^>]*>.*?<span[^>]*class="[^"]*number[^"]*"[^>]*>([^<]+)</span>'
            for m in re.finditer(alt_ep_pattern, html, re.DOTALL):
                episodes.append({'id': m.group(1), 'name': m.group(2).strip()})

        # 3. 组装数据结构
        play_froms, play_urls = [], []
        for line in lines:
            if not episodes: 
                continue
            play_froms.append(line['name'])
            ep_list = [f"{ep['name']}${line['id']}-{vod_id}-{ep['id']}" for ep in episodes]
            play_urls.append('#'.join(ep_list))
        
        # 4. 兜底逻辑（无线路时用默认线路）
        if not play_froms and episodes:
            play_froms.append('默认线路')
            ep_list = [f"{ep['name']}${vod_id}-{ep['id']}" for ep in episodes]
            play_urls.append('#'.join(ep_list))

        return {
            'vod_play_from': '$$$'.join(play_froms),
            'vod_play_url': '$$$'.join(play_urls)
        }

    # ----------------------------------------------
    # 公共接口实现
    # ----------------------------------------------

    def init(self, cfg):
        """初始化会话"""
        self.session = requests.Session()
        self.session.headers.update(self.headers)

    def homeContent(self, filter):
        """首页分类数据"""
        return {
            'class': [
                {'type_id': '1', 'type_name': '电影'},
                {'type_id': '2', 'type_name': '电视剧'},
                {'type_id': '4', 'type_name': '动漫'},
                {'type_id': '3', 'type_name': '综艺'},
                {'type_id': '8', 'type_name': '短剧'},
                {'type_id': '6', 'type_name': '纪录片'}
            ],
            'filters': {}
        }

    def homeVideoContent(self):
        """首页视频（空实现）"""
        return {'list': []}

    def categoryContent(self, tid, pg, filter, ext):
        """分类页数据抓取"""
        page = max(1, int(pg or 1))
        url = f"{self.home_url}/channel?page={page}&cat_id={tid}&order=new&page_size=32"
        resp = self._make_request(url, headers=self.headers)
        if not resp: 
            return {'list': [], 'page': page, 'pagecount': 0}
        
        items = re.findall(r'<div class="item-wrap[\s\S]*?</div>\s*</div>', resp.text)
        video_list = []
        
        for item in items:
            # 标题 & ID
            title = re.search(r'<a[^>]*title="([^"]+)"', item)
            href = re.search(r'<a[^>]*href="(/movie/detail/[^"]+)"', item)  # 严格过滤路径
            if not (title and href): 
                continue
            
            # 封面图
            pic = re.search(r'data-src="([^"]+)"', item)
            pic_url = pic.group(1) if pic else self._get_placeholder_pic(title.group(1))
            
            # 备注
            remark = re.search(r'<div class="category[^>]*>([^<]+)</div>', item)
            
            video_list.append({
                'vod_id': href.group(1),
                'vod_name': title.group(1),
                'vod_pic': pic_url,
                'vod_remarks': remark.group(1).strip() if remark else ''
            })
        
        return {'list': video_list, 'page': page, 'pagecount': 99}  # 模拟分页上限

    def detailContent(self, did):
        """详情页数据抓取"""
        vod_id = did[0] if isinstance(did, list) else did
        if not vod_id or not vod_id.startswith('/movie/detail/'):
            return {'list': []}
        
        url = f"{self.home_url}{vod_id}"
        resp = self._make_request(url, headers=self.headers)
        if not resp: 
            return {'list': []}
        
        html = resp.text
        # 标题
        name = re.search(r'<h1 class="title[^>]*>([^<]+)</h1>', html)
        name = name.group(1).strip() if name else '未知影片'
        
        # 封面图
        pic = re.search(r'class="thumb[^>]*data-src="([^"]+)"', html)
        pic_url = pic.group(1) if pic else self._get_placeholder_pic(name)
        
        # 简介
        desc = re.search(r'<div class="desc[^>]*>([\s\S]*?)</div>', html)
        content = re.sub(r'<[^>]+>', '', desc.group(1)).strip() if desc else ''
        
        # 类型
        type_name = re.search(r'类型：<a[^>]*>([^<]+)</a>', html)
        type_name = type_name.group(1).strip() if type_name else ''
        
        # 线路&集数
        play_data = self._extract_lines_and_episodes(html, vod_id)
        
        return {'list': [{
            'vod_id': vod_id,
            'vod_name': name,
            'vod_pic': pic_url,
            'vod_content': content,
            'type_name': type_name,
            'vod_play_from': play_data['vod_play_from'],
            'vod_play_url': play_data['vod_play_url'],
            'vod_remarks': '',  # 备注在分类页已用，详情页留空
            'vod_year': '',
            'vod_actor': '',
            'vod_director': ''
        }]}

    def searchContent(self, key, quick, pg):
        """搜索功能（同分类逻辑）"""
        page = int(pg) if pg else 1
        url = f"{self.home_url}/channel?page={page}&keywords={key}&page_size=32&order=new"
        resp = self._make_request(url)
        if not resp:
            return {'list': []}
        html = resp.text

        items = re.findall(r'<div class="item-wrap[\s\S]*?</div>\s*</div>', html)
        video_list = []
        for item_html in items:
            title_match = re.search(r'<a[^>]*title="([^"]+)"', item_html)
            href_match = re.search(r'<a[^>]*href="([^"]+)"', item_html)
            if not title_match or not href_match:
                continue
            name = title_match.group(1)
            vid = href_match.group(1)
            if not vid.startswith('/movie/detail/') or '://' in vid:
                continue

            pic = ''
            pic_match = re.search(r'data-src="([^"]+)"', item_html)
            if pic_match:
                pic = pic_match.group(1)
            if not pic:
                pic = self._get_placeholder_pic(name)

            remark = ''
            remark_match = re.search(r'<div class="category[^>]*>([^<]+)</div>', item_html)
            if remark_match:
                remark = remark_match.group(1).strip()

            video_list.append({
                'vod_id': vid,
                'vod_name': name,
                'vod_pic': pic,
                'vod_remarks': remark
            })

        return {'list': video_list, 'page': page, 'pagecount': 99}

    def playerContent(self, flag, pid, vipFlags):
        """
        播放器核心逻辑
        pid格式: lineId-vodId-epId 或 vodId-epId
        """
        # 1. 解析pid
        parts = pid.split('-')
        if len(parts) >= 3:
            line_id, vod_id, ep_id = parts[0], parts[1], parts[2]
        elif len(parts) == 2:
            line_id, vod_id, ep_id = 'default', parts[0], parts[1]
        else:
            # 无法解析时，返回一个错误状态，而不是抛出异常
            return {'parse': 1, 'url': f"{self.home_url}/error", 'header': self.headers}

        # 2. 构造请求
        detail_url = f"{self.home_url}{vod_id}"
        cookies = {'_did': self._generate_cookie()}
        data = {'link_id': ep_id, 'is_switch': '1'}
        
        try:
            resp = self.session.post(
                detail_url,
                data=data,
                cookies=cookies,
                headers=self.headers,
                timeout=self.timeout
            )
            resp.raise_for_status()
            json_data = resp.json()
            
            # 3. 提取m3u8
            play_links = json_data.get('data', {}).get('play_links', [])
            m3u8_url = None
            
            if line_id == 'default' and play_links:
                m3u8_url = play_links[0].get('m3u8_url')
            else:
                for link in play_links:
                    if str(link.get('id')) == line_id:
                        m3u8_url = link.get('m3u8_url')
                        break
            
            # 4. 修正URL
            if m3u8_url:
                # 核心修复：如果是相对路径，确保它以 / 开头，然后与 home_url 拼接
                if not m3u8_url.startswith('http'):
                    if not m3u8_url.startswith('/'):
                        m3u8_url = '/' + m3u8_url
                    full_url = self.home_url + m3u8_url
                else:
                    full_url = m3u8_url
                return {
                    'parse': 0,
                    'url': full_url,
                    'header': self.headers
                }
            else:
                # 没拿到m3u8，尝试嗅探
                return {'parse': 1, 'url': detail_url, 'header': self.headers}
                
        except Exception as e:
            print(f"播放器错误: {str(e)}")
            # 出错时也返回一个有效的字典，让播放器去试试嗅探
            return {'parse': 1, 'url': detail_url, 'header': self.headers}

    def destroy(self):
        """资源清理"""
        if self.session:
            self.session.close()
            self.session = None
